### ISSUE TITLE
Spikes in values read from the tlv493d magnetometer

### ISSUE THREAD

**Issue by lubarb** [adafruit/Adafruit_CircuitPython_TLV493D#9]
Reading values from the tlv493d with no magnet near, most of the values returned are in the range of roughly +- .3 units, consistent with bit and noise levels from the spec sheet. However about 1-2% of the readings are roughly +- 1.5 units.

![Screenshot 2021-03-16 095848](https://user-images.githubusercontent.com/15658271/111323950-b3f2e100-8640-11eb-98c3-2e7a223a733c.png)

Logged with the following code:

```
import board
import math
import time
import wifi
import socketpool
import ssl
import json
import busio
import adafruit_tlv493d
import adafruit_minimqtt.adafruit_minimqtt as MQTT

DOOR_MQTT_TOPIC = "dryertest/door"

# Hardware I2C setup:

i2c = busio.I2C(board.SCL, board.SDA)
tlv = adafruit_tlv493d.TLV493D(i2c)

print("starting up")

# Add a secrets.py to your filesystem that has a dictionary called secrets with "ssid" and
# "password" keys with your WiFi credentials. DO NOT share that file or commit it into Git or other
# source control.
# pylint: disable=no-name-in-module,wrong-import-order
try:
    from secrets import secrets
except ImportError:
    print("WiFi secrets are kept in secrets.py, please add them there!")
    raise

wifi.radio.connect(secrets["ssid"], secrets["password"])
print("Connected to %s!" % secrets["ssid"])

# Create a socket pool
pool = socketpool.SocketPool(wifi.radio)

# Set up a MiniMQTT Client
mqtt_client = MQTT.MQTT(
    broker=secrets["mqtt_broker"],
    port=secrets["mqtt_port"],
    username=secrets["mqtt_username"],
    password=secrets["mqtt_password"],
    socket_pool=pool,
    ssl_context=ssl.create_default_context(),
)

print("Attempting to connect to %s" % mqtt_client.broker)
mqtt_client.connect()

readcount = 0
badvalue = False

while True:
    door = tlv.magnetic
    door_output = {
    "x" : door[0],
    "y" : door[1],
    "z" : door[2],
    }

    readcount += 1

    if (abs(door[0]) > 1.0) :
        print("Outlier value (X) seen after ", readcount, " reads")
        badvalue = True
    if (abs(door[1]) > 1.0) :
        print("Outlier value (Y) seen after ", readcount, " reads")
        badvalue = True
    if (abs(door[2]) > 1.0) :
        print("Outlier value (Z) seen after ", readcount, " reads")
        badvalue = True
    if badvalue :
        badvalue = False
        readcount = 0

    print('.', end='')
    
    mqtt_client.publish(DOOR_MQTT_TOPIC, json.dumps(door_output))

    time.sleep(1)

```
This seems to be consistent across multiple sensor boards connected to a Magtag and to several Metro ESP32-S2 boards with several Stemma cables.

My most recent test is with a Magtag using Adafruit CircuitPython 6.2.0-beta.3 on 2021-03-04; Adafruit MagTag with ESP32S2 and libraries from 3/15/21 with the exception of a slightly edited adafruit_minimqtt.py (the library version crashes in the absence of logging, a pull request hadn't made it into the library release). I've also seen the behavior with the 6.1.0 release of CircuitPython and earlier libraries.

Also, in looking at the adafruit_tlv493d.py library, the __init__ function sets up values to be written for control into a buffer before actually writing the values out over i2c.

```
        # setup MASTERCONTROLLEDMODE which takes a measurement for every read
        self._set_write_key("PARITY", 1)
        self._set_write_key("PARITY", 1)
        self._set_write_key("LOWPOWER", 1)
        self._set_write_key("LP_PERIOD", 1)
        self._write_i2c()

```
Looking at the code for _set_write_key, entering the PARITY value into the buffer twice is redundant. I can't figure out from the chip documentation whether this was just an inadvertent duplication or whether perhaps setting another different control bit was intended.

**Comment by ladyada** [comment]
could be the wifi signal is inducing a signal. we recommend taking multiple readings and using a running average

**Comment by lubarb** [comment]
I may have identified a fix for my problem.

Replace:
```
        self._set_write_key("PARITY", 1)
        self._set_write_key("PARITY", 1)
        self._set_write_key("LOWPOWER", 1)
        self._set_write_key("LP_PERIOD", 1)

```
with:
```
        self._set_write_key("PARITY", 1)
        self._set_write_key("FAST", 1)
        self._set_write_key("LOWPOWER", 1)

```
What I did (ignoring MANY red herrings I chased):

Where the source is setting up the sensor, it has the comment "setup MASTERCONTROLLEDMODE which takes a measurement for every read". The chip user manual section 5.5 also specifies that this mode performs one measurement cycle per read. 

In section 7.2.1.1 the description of the Temp register FRM field indicates that the frame counter should increment by one with each complete read. I extracted the frame counter for each measurement and found that the frame counter values seemed to be essentially random. From this, I suspected that the conversions were not running in sync with the reads being returned and that the chip may have been running in another mode.

Looking back at section 5.5 it says "Settings: fastmode = 1, lp_mode = 1, int_out = 0/1 (byte settings: 00, x7, xx, xx, keep certain bits)". I replaced the setup in the code with this setup.

After this change the frame counter value incremented as expected.

With no magnet near, I can read 10,000+ measurements without any spikes (and it does still respond to the presence of a magnet).

But:
I find the chip documentation obscure to say the least. I haven't a clue to what may have been happening before and whether/why it might have been causing the spikes I got. But I now get values that look like (10x as many samples as in the original screenshot): 
![Screenshot 2021-03-19 123013](https://user-images.githubusercontent.com/15658271/111826312-e9a4fd80-88be-11eb-80bc-598bfc498a85.png)
